File: src/node/forensic_helpers.js
/*
    Forensic Helpers (Node.js)
    Purpose:
        Toggle‑driven, JSON‑formatted forensic logging with correlation IDs,
        assumption checks, and scoped timing.

    Exports:
        - forensicLog(message: string, startTime: Date, level: string = "INFO")
        - forensicCheck(condition: boolean, message: string)
        - forensicScope(name: string, fn: Function)  // cross‑language parity

    Scope Options:
        1. Cross‑language parity:
            forensicScope("MyBlock", () => {
                // ... logic ...
            });

        2. Idiomatic Node.js try/finally:
            const start = new Date();
            forensicLog("MyBlock start", start);
            try {
                // ... logic ...
            } finally {
                forensicLog("MyBlock end", start);
            }

    Example:
        // Enable logging
        export const forensicOn = true;

        // Log a checkpoint
        forensicLog("Starting process", forensicStart);

        // Validate assumption
        forensicCheck(items.length > 0, "No items found");

        // Scoped timing (parity)
        forensicScope("LoadData", () => {
            loadData();
        });

        // Scoped timing (idiomatic)
        const start = new Date();
        forensicLog("ProcessData start", start);
        try {
            processData();
        } finally {
            forensicLog("ProcessData end", start);
        }
*/

import crypto from "crypto";

export const forensicOn = true;
export const correlationId = crypto.randomUUID();
export const forensicStart = new Date();

export function forensicLog(message, startTime, level = "INFO") {
    if (!forensicOn) return;
    const now = new Date();
    const elapsedMs = now - startTime;
    console.log(JSON.stringify({
        ts: now.toISOString(),
        elapsed_ms: elapsedMs,
        corr_id: correlationId,
        level,
        msg: message
    }));
}

export function forensicCheck(condition, message) {
    if (forensicOn && !condition) {
        forensicLog(message, forensicStart, "WARN");
    }
}

export function forensicScope(name, fn) {
    const start = new Date();
    forensicLog(`${name} start`, start, "INFO");
    try {
        fn();
    } finally {
        forensicLog(`${name} end`, start, "INFO");
    }
}

File: src/go/forensic_helpers.go
//------------------------------------------------------------------------------
// Forensic Helpers (Go)
// Purpose:
//     Toggle‑driven, JSON‑formatted forensic logging with correlation IDs,
//     assumption checks, and scoped timing.
//
// Exports:
//     - forensicLog(message string, start time.Time, level string)
//     - forensicCheck(condition bool, message string)
//     - forensicScope(name string, fn func())       // cross‑language parity
//     - startScope(name string) (endFn func())      // idiomatic defer pattern
//
// Scope Options:
//     1. Cross‑language parity:
//         forensicScope("MyBlock", func() {
//             // ... logic ...
//         })
//
//     2. Idiomatic Go defer:
//         end := startScope("MyBlock")
//         defer end()
//         // ... logic ...
//
// Example:
//     package main
//
//     import (
//         "time"
//         "myorg/forensic"
//     )
//
//     func main() {
//         forensic.ForensicOn = true
//
//         // Log a checkpoint
//         forensic.ForensicLog("Starting process", forensic.ForensicStart, "INFO")
//
//         // Validate assumption
//         forensic.ForensicCheck(len(items) > 0, "No items found")
//
//         // Scoped timing (parity)
//         forensic.ForensicScope("LoadData", func() {
//             loadData()
//         })
//
//         // Scoped timing (idiomatic)
//         end := forensic.StartScope("ProcessData")
//         defer end()
//         processData()
//     }
//------------------------------------------------------------------------------

package forensic

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

var ForensicOn = true
var correlationId = uuid.NewString()
var ForensicStart = time.Now().UTC()

type logEntry struct {
	Ts        string `json:"ts"`
	ElapsedMs int64  `json:"elapsed_ms"`
	CorrID    string `json:"corr_id"`
	Level     string `json:"level"`
	Msg       string `json:"msg"`
}

func ForensicLog(message string, start time.Time, level string) {
	if !ForensicOn {
		return
	}
	now := time.Now().UTC()
	elapsed := now.Sub(start).Milliseconds()
	entry := logEntry{
		Ts:        now.Format(time.RFC3339Nano),
		ElapsedMs: elapsed,
		CorrID:    correlationId,
		Level:     level,
		Msg:       message,
	}
	b, _ := json.Marshal(entry)
	fmt.Println(string(b))
}

func ForensicCheck(condition bool, message string) {
	if ForensicOn && !condition {
		ForensicLog(message, ForensicStart, "WARN")
	}
}

// Cross‑language parity wrapper
func ForensicScope(name string, fn func()) {
	start := time.Now().UTC()
	ForensicLog(fmt.Sprintf("%s start", name), start, "INFO")
	defer ForensicLog(fmt.Sprintf("%s end", name), start, "INFO")
	fn()
}

// Idiomatic Go start/defer end pattern
func StartScope(name string) func() {
	start := time.Now().UTC()
	ForensicLog(fmt.Sprintf("%s start", name), start, "INFO")
	return func() {
		ForensicLog(fmt.Sprintf("%s end", name), start, "INFO")
	}
}
